src/core/utils/spec-preprocess.ts

```diff
@@ -187,8 +188,12 @@ export function traverseToFixSpecDownstream(spec: GoslingSpec | SingleView, pare
         const linkID = uuid.v4();
         tracks.forEach((track, i, array) => {
             // If size not defined, set default ones
-            if (!track.width) track.width = DEFAULT_TRACK_WIDTH_LINEAR;
-            if (!track.height) track.height = DEFAULT_TRACK_HEIGHT_LINEAR;
+            if (!track.width) {
+                track.width = Is2DTrack(track) ? DEFAULT_TRACK_SIZE_2D : DEFAULT_TRACK_WIDTH_LINEAR;
+            }
+            if (!track.height) {
+                track.height = Is2DTrack(track) ? DEFAULT_TRACK_SIZE_2D : DEFAULT_TRACK_HEIGHT_LINEAR;
+            }
```

Comment:
Ah I see, so here, we populate the spec with defined if they are missing. This is why a function that mutates the spec is unfortunate; we cannot capture that we've changed the type very easily. A function,

```typescript 
defineTrackSizes(spec); // mutates track sizes if missing
spec // spec still has same type (tracks could be missing)
```

vs

```typescript
const specWithCompleteTrackSizes = defineTrackSizes(spec); // does not mutate spec
```

Reply:
I also wonder if we should handle the case where it's a 2DTrack but only width _or_ height is defined. In that case, we probably want to use the defined dimension to infer the other? Actually not sure...

Reply:
`traverseToFixSpecDownstream` expands a spec by adding default values, and after calling this function, many optional properties are non-null. I had an idea of having another type of `GoslingSpec` that we internally use (something like `_GoslingSpec`), such as making some optional properties required and adding internally used properties (e.g., `startAngle`). But, I wasn't sure what would be the best and cleanest approach to create a new schema given that the `GoslingSpec` is fairly complex and deeply nested. At maximum, we could create a new type per each type we already have (e.g., `_SingleTrack`, `_Track`, `_CommonViewDef`, ..., `_GoslingSpec`).

Reply:
> I also wonder if we should handle the case where it's a 2DTrack but only width or height is defined. In that case, we probably want to use the defined dimension to infer the other? Actually not sure...

I thought the same but was not sure either. It might be better if we expect that users will most frequently use matrices with the same `width` and `height` and it does not confuse users. Being able to specify only one of them might be a bit handier.

Reply:
> `traverseToFixSpecDownstream` expands a spec by adding default values, and after calling this function, many optional properties are non-null. 

We might be able to write a custom type helper for this. e.g,. `WithDefinedTracks<GoslingSpec>` and then just assert that the tracks are defined. But the `GoslingSpec` type is complicated recursive type so it might be challenging. I'd try to avoid maintaining `_GoslingSpec`, etc.